src/ucai/core/databricks.py

```diff
@@ -168,6 +221,14 @@ def _validate_warehouse_type(self):
                 "https://docs.databricks.com/en/admin/sql/serverless.html#enable-serverless-sql-warehouses."
             )
 
+    def refresh_client_and_session(self):
+        _logger.info("Refreshing Databricks client and Spark session due to session expiration.")
+        self.client = get_default_databricks_workspace_client(profile=self.profile)
+        self.stop_spark_session()
+        self.set_default_spark_session()
+        self.spark = _try_get_spark_session_in_dbr()
```

Comment:
I had a simple test, this logic seems broken if in databricks notebook, since we can't directly reinitialize the spark session (which requires some connection string) in notebook.
Can we check if we're in dbr first, and if so we do nothing to the spark session (or we can investigate to see how to refresh the cluster); and if we're not in dbr, we can stop + set again.

Reply:
Yep - confirmed. It's completely broken in serverless (it also makes a fun time of requiring a termination / restart command if this method is called!). Added a check for interactive notebook validation, removal of spark session termination and relying on a refresh of the client instance for the WorkspaceClient and simply refreshing the reference to the notebook REPL SparkSession obj.